client/src/graphs/templates/useTemplate.ts

```diff
@@ -0,0 +1,96 @@
+import type { Graph } from "@graph/types";
+import type { GraphTemplate } from "./types";
+import { computed, ref } from "vue";
+import {
+  centerNodesOnOriginCoordinates,
+  getAverageCoordinatesOfGraphNodes,
+  getGraphNodesBoundingBox,
+} from "./helpers";
+import { generateId } from "@utils/id";
+import { useLocalStorage } from "@vueuse/core";
+import { createImageFromCanvasRegion } from "./snapshot";
+import { products } from "@utils/product";
+
+export const useTemplate = (graph: Graph) => {
+  const userTemplates = useLocalStorage<GraphTemplate[]>(
+    "user-graph-templates",
+    []
+  );
+  const productTemplates = ref<GraphTemplate[]>(
+    products.flatMap((p) => p.templates ?? [])
+  );
+
+  const templates = computed(() => [
+    ...userTemplates.value, // user templates first so easier to find
+    ...productTemplates.value,
+  ]);
+
+  const addCurrentGraphAsTemplate = (
+    templateDetails: Pick<GraphTemplate, "title" | "description" | "thumbnail">
+  ) => {
+    const { nodes, edges } = graph;
+
+    const nodeRadius = graph.baseTheme.value.nodeSize;
+    const nodeBorderWidth = graph.baseTheme.value.nodeBorderWidth;
+    const boundingBox = getGraphNodesBoundingBox(
+      nodes.value,
+      nodeRadius + nodeBorderWidth / 2
+    );
+
+    if (!graph.canvas.value) return;
```

Comment:
its kinda weird to do this here. either make it guard at the top or not at all, we already do a bunch of canvas computations. i would also just edit the param of image to take it a ref that could be undefined. GetCtx does the checks for you